components/ActionSheet/src/private/MDCActionSheetHeaderView.m

```diff
@@ -0,0 +1,241 @@
+static const CGFloat TopStandardPadding = 16.f;
+static const CGFloat TrailingPadding = 16.f;
+static const CGFloat TitleOnlyPadding = 18.f;
+static const CGFloat MiddlePadding = 8.f;
+
+@interface MDCActionSheetHeaderView ()
+
+@property(nonatomic, nonnull, strong) UIScrollView *scrollView;
+
+@end
+
+@implementation MDCActionSheetHeaderView {
+  UILabel *_titleLabel;
+  UILabel *_messageLabel;
+}
+
+@synthesize mdc_adjustsFontForContentSizeCategory;
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    self.scrollView = [[UIScrollView alloc] initWithFrame:CGRectZero];
+    _titleLabel = [[UILabel alloc] initWithFrame:CGRectZero];
+    _messageLabel = [[UILabel alloc] initWithFrame:CGRectZero];
+    [self addSubview:self.scrollView];
+    [self.scrollView addSubview:_titleLabel];
+    _titleLabel.font = [MDCTypography subheadFont];
+    _titleLabel.numberOfLines = 0;
+    _titleLabel.lineBreakMode = NSLineBreakByTruncatingMiddle;
+
+    [self.scrollView addSubview:_messageLabel];
+    _messageLabel.font = [MDCTypography body1Font];
+    _messageLabel.numberOfLines = 2;
+    _messageLabel.lineBreakMode = NSLineBreakByWordWrapping;
+    _messageLabel.alpha = MessageLabelAlpha;
+  }
+  return self;
+}
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
+
+  if (self.message == nil || [self.message  isEqualToString:@""]) {
+    _titleLabel.alpha = MessageLabelAlpha;
+  } else {
+    _titleLabel.alpha = TitleLabelAlpha;
+  }
+  CGSize boundsSize = CGRectInfinite.size;
+  boundsSize.width = [self accomodateSafeAreaInWidth:CGRectGetWidth(self.bounds)];
+
+  CGSize contentSize = [self calculateContentSizeThatFitsWidth:boundsSize.width];
+  self.scrollView.contentSize = contentSize;
+
+  boundsSize.width = boundsSize.width - LeadingPadding - TrailingPadding;
+  CGSize titleSize = [_titleLabel sizeThatFits:boundsSize];
+  titleSize.width = boundsSize.width;
+  CGSize messageSize = [_messageLabel sizeThatFits:boundsSize];
+  messageSize.width = boundsSize.width;
+  boundsSize.width = boundsSize.width + LeadingPadding + TrailingPadding;
+
+  CGRect titleFrame = CGRectMake(LeadingPadding, TopStandardPadding,
+                                 titleSize.width, titleSize.height);
+  CGRect messageFrame = CGRectMake(LeadingPadding, CGRectGetMaxY(titleFrame) + MiddlePadding,
+                                   messageSize.width, messageSize.height);
+  _titleLabel.frame = titleFrame;
+  _messageLabel.frame = messageFrame;
+
+  CGRect scrollViewRect = CGRectZero;
+  scrollViewRect.size = self.scrollView.contentSize;
+#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
+  if (@available(iOS 11.0, *)) {
+    scrollViewRect.origin.x = scrollViewRect.origin.x + self.safeAreaInsets.left;
+  }
+#endif
+
+  self.scrollView.frame = scrollViewRect;
+  CGRect originalFrame = self.frame;
+  originalFrame.size.height = scrollViewRect.size.height;
+  self.frame = originalFrame;
+}
```

Comment:
I think you may be able to reduce some replicated logic between layoutSubviews and sizeThatFits: by following the pattern outlined here:

https://gist.github.com/jverkoey/5288470

This pattern lets you reuse local state variables for each pass, while only affecting actual layout when you need to.

Reply:
Done